### PR TITLE
improving is_perfect and is_mersenne_prime methods

### LINKED ISSUE #16144: shortcut for is_perfect
Since "Every even perfect number ends in 6 or 28 base 10" that would be a good first check for the newly added `is_perfect` function in ntheory: 
```python
last2 = n%100
if last2 != 28 and last2%10 != 6:
    return False
```

### PR DESCRIPTION
**Pull request by mohitacecode**
#### References to other Issues or PRs

Fixes #16144 
#### Brief description of what is fixed or changed
Perfomance of `is_perfect` and `is_mersenne_prime` is improved as suggested in pr(https://github.com/sympy/sympy/pull/16145#issuecomment-471144618)

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/ntheory/factor_.py
Comment (ethankward) [inline sympy/ntheory/factor_.py]: This takes a large amount of memory- these are massive numbers. I think you should find a better way to cache this.

**Review by ethankward — CHANGES_REQUESTED** [review]

File: sympy/ntheory/factor_.py
Comment (mohitacecode) [reply sympy/ntheory/factor_.py]: It was discussed that we should do whatever faster is here. (https://github.com/sympy/sympy/pull/16145#issuecomment-470649409).

File: sympy/ntheory/factor_.py
Comment (mohitacecode) [inline sympy/ntheory/factor_.py]: does it mean `integer_nthroot` because we dont have any method called `integer_root`.

File: sympy/ntheory/factor_.py
Comment (asmeurer) [reply sympy/ntheory/factor_.py]: We could recompute them each time as well. It depends on whether we consider this to be too much memory usage.

Comment (mohitacecode) [reply sympy/ntheory/factor_.py]: If we are in favour of saving memory then we can use something like `python generators` and can generate the term whenever we required.(Will save lot of memory)
But all this process will take too much time and IMO we should give priority to time as no. of terms are limited here.

File: sympy/ntheory/factor_.py
Comment (mohitacecode) [reply sympy/ntheory/factor_.py]: what about this. @asmeurer

File: sympy/ntheory/factor_.py
Comment (smichr) [reply sympy/ntheory/factor_.py]: An even number is perfect if it can be expressed as two factors: one a power of two and the other being one less than a related power of two. Except for the first 23 Mersenne prime powers it would be much faster to check the exponent:
```python
def _isperfect(n):
    if n%2 == 0:
        i = multiplicity(2, n)
        o = n >> i
        return multiplicity(2, o + 1) == i + 1
```
So I would suggest precalculating the first 23 perfect numbers and then for a given number do
```python
PERFECT = [2**(i-1)*(2**i-1) for i in MERSENNE_PRIME_EXPONENTS[:23]]
def _isperfect(n):
    global perfect
    if n % 2 == 0:
        while n > PERFECT[-1] and len(PERFECT) < len(MERSENNE_PRIME_EXPONENTS):
            i = 2**(MERSENNE_PRIME_EXPONENTS[len(PERFECT)] - 1)
            PERFECT.append(i*(2*i - 1))
        if n <= PERFECT[-1]:
            return n in PERFECT
    else:
        i,_ = integer_log(n, 2)
        if i < exponent below which all odds are known to be not perfect:
            return False
    # handle new analysis for odd or even
```

Comment (ethankward) [reply sympy/ntheory/factor_.py]: It takes 5 seconds to initialize `PERFECT = [2**(i - 1)*(2**i - 1) for i in MERSENNE_PRIME_EXPONENTS]` on a powerful computer.I don't know what you mean by "priority to time" but that is not fast.

File: sympy/ntheory/factor_.py
Comment (smichr) [inline sympy/ntheory/factor_.py]:
```suggestion
        odd = [(b, e) for b, e in facs.items() if e%2]  # only 1 odd exponent
```

File: sympy/ntheory/factor_.py
Comment (smichr) [reply sympy/ntheory/factor_.py]: I rewrote this and committed.

File: sympy/ntheory/factor_.py
Comment (asmeurer) [reply sympy/ntheory/factor_.py]: I think that's the point. If we cache the values that 5 seconds only needs to be run once. But it requires storing the numbers in memory. 

Personally I think that

- the 100 or so MB that this takes isn't so bad on a modern computer. But neither is 5 seconds.
- very few people will actually call these functions. I don't think they would be used in other places in SymPy (or am I wrong?). So it doesn't actually matter that much.

**Comment by mohitacecode** [comment]
@smichr Have you done all the changes? sorry I was working on different pr.

File: sympy/ntheory/factor_.py
Comment (ethankward) [reply sympy/ntheory/factor_.py]: With the incremental caching it's much better.

**Comment by smichr** [comment]
> @smichr Have you done all the changes? sorry I was working on different pr.

yes, see what you think.

**Comment by smichr** [comment]
This needs a final ok from @ethankward

**Review by smichr — APPROVED** [review]
I see that @ethankward already marked the changes as acceptable so I think this is ready to go. Will wait for any other feedback, though.

**Comment by smichr** [comment]
Nope...looks like once someone has reviewed and requested changes they have to sign off on them. So this is waiting for @ethankward

**Review by ethankward — APPROVED** [review]

**Pull request merged**
